### PR TITLE
[DTensor] Add layer_norm decomposition handler to avoid forced redistribution

### PR DESCRIPTION
**Pull request by RohitRathore1**
When input is sharded on normalized dimensions, decompose layer_norm into local sum reductions followed by a single all-reduce instead of forcing full tensor redistribution. Reduces communication from O(tensor_size) to O(outer_dims_size).

So, basically added custom handler for `native_layer_norm` that decomposes the operation when input is sharded on normalized dimensions. Instead of forcing redistribution to Replicate (all-gather), we:

1. Compute local `sum(x)` and `sum(x²)` → Partial placement
2. Stack and all-reduce in a single collective
3. Compute mean/variance from global sums
4. Normalize locally (input stays sharded)

FIXES #174276

### CODE REVIEW

File: torch/distributed/tensor/_nonlinear_redux.py [hunk 1]
```diff
@@ -295,3 +295,94 @@ def minmax_dim_handler(
...
+    )
+    stacked = stacked.redistribute(placements=new_placements)
+    sum_x = stacked[0]
+    sum_x2 = stacked[1]
+
+    mean = sum_x / N
+    var = sum_x2 / N - mean * mean
+    rstd = torch.rsqrt(var + eps)
+
+    out = (input_tensor - mean) * rstd
```
Comment (zpcore) [inline torch/distributed/tensor/_nonlinear_redux.py:381]: With those manual implementation, are we going to lose the opportunity for kernel optimization?

File: torch/distributed/tensor/_nonlinear_redux.py [hunk 2]
```diff
@@ -295,3 +295,94 @@ def minmax_dim_handler(
...
+
+    # compute local sums (these will have Partial placement)
+    sum_x = input_tensor.sum(dim=reduction_dims, keepdim=True)
+    sum_x2 = (input_tensor * input_tensor).sum(dim=reduction_dims, keepdim=True)
+
+    # all-reduce to get global sums
+    # stack to do a single all-reduce, then unstack
+    stacked = cast(dtensor.DTensor, torch.stack([sum_x, sum_x2], dim=0))
+    new_placements = tuple(
+        Replicate() if p.is_partial() else p for p in stacked.placements
```
Comment (zpcore) [inline torch/distributed/tensor/_nonlinear_redux.py:371]: Will
```
Replicate() if not p.is_replicate() else p for p in stacked.placements
```
safer?

File: torch/distributed/tensor/_nonlinear_redux.py [hunk 3]
```diff
@@ -295,3 +295,94 @@ def minmax_dim_handler(
...
+    )
+    stacked = stacked.redistribute(placements=new_placements)
```
Comment (zpcore) [inline torch/distributed/tensor/_nonlinear_redux.py:373]: This is interesting that you call sum op over the `input_tensor` Dtensor and trigger redistribute. I've never seen such use case for custom handler, but this seems work?

**Review by wconstab — COMMENTED** [review]
This is a clever idea. It is also not very aligned with the DTensor design. I'm not sure it's worth doing this optimization at the expense of making the system behave differently from expected.

(What do i mean?)

The design principle is that DTensor always does 
- sharding prop to list the operator's supported shardings
- redistribution to min cost strategy (if needed)
- run the local op

This PR clearly violates that by doing extra/different comms 'inside' the decomposed op.

I'll let others weigh in but i'd probably lean towards not doing this and instead writing user code that does this.

**Comment by pianpwk** [comment]
> This is a clever idea. It is also not very aligned with the DTensor design. I'm not sure it's worth doing this optimization at the expense of making the system behave differently from expected.

This is exactly what we were gonna do for vector_norm (with NormPartial removal) though, I mentioned that PR in the issue linked. I think it could be reasonable to have all norm ops follow this cheaper approach if we eventually deprecate? cc @tianyu-l

**Comment by wconstab** [comment]
my question would be around whether we have dtensor do this automatically or if we put code into torchtitan that does the 'smart thing'.  the tradeoff is whether this optimization is more automatic but perhaps surprising, or simpler but requires more manual work.

File: torch/distributed/tensor/_nonlinear_redux.py (hunk 2 above)
Comment (RohitRathore1) [reply torch/distributed/tensor/_nonlinear_redux.py:371]: using `not p.is_replicate()` would convert `Shar -> Replicate`, triggering unnecessary communication on non-reduced dimensions, right?

File: torch/distributed/tensor/_nonlinear_redux.py (hunk 1 above)
Comment (RohitRathore1) [reply torch/distributed/tensor/_nonlinear_redux.py:381]: Yes, but communication cost dominates in distributed case. I think communication reduction matters more than kernel efficiency when data is sharded across gpus.

File: torch/distributed/tensor/_nonlinear_redux.py (hunk 3 above)
Comment (RohitRathore1) [reply torch/distributed/tensor/_nonlinear_redux.py:373]: Yes, this works. DTensor ops inside handlers trigger normal dispatch. Same pattern in #172620 (vector_norm_handler). The sum() dispatches through DTensor, produces Partial placement, then we redistribute explicitly.

File: torch/distributed/tensor/_nonlinear_redux.py (hunk 2 above)
Comment (zpcore) [reply torch/distributed/tensor/_nonlinear_redux.py:371]: I mean in this case, `p` may be Partial('max') (or not possible?).
